File: shared/protocol.py
"""
MacWinControl - Shared Protocol
Defines message types for bidirectional communication
"""

from dataclasses import dataclass, asdict
from typing import Optional, List, Dict, Any
from enum import Enum
import json


class MessageType(Enum):
    # Connection
    HELLO = "hello"
    CONNECTED = "connected"
    DISCONNECT = "disconnect"
    PING = "ping"
    PONG = "pong"
    
    # Discovery
    DISCOVER = "discover"
    ANNOUNCE = "announce"
    
    # Mode switching
    MODE_SWITCH = "mode_switch"
    
    # Input events
    MOUSE_MOVE = "mouse_move"
    MOUSE_BUTTON = "mouse_button"
    MOUSE_SCROLL = "mouse_scroll"
    KEY_DOWN = "key_down"
    KEY_UP = "key_up"
    
    # Clipboard
    CLIPBOARD_SYNC = "clipboard_sync"
    CLIPBOARD_REQUEST = "clipboard_request"
    
    # Screen config
    SCREEN_INFO = "screen_info"
    SCREEN_LAYOUT = "screen_layout"
    
    # Settings
    SETTINGS_UPDATE = "settings_update"


@dataclass
class ScreenInfo:
    """Information about a screen/monitor"""
    id: str
    name: str
    width: int
    height: int
    x: int  # Position in virtual desktop
    y: int
    is_primary: bool = False
    scale: float = 1.0


@dataclass
class ComputerInfo:
    """Information about a computer"""
    id: str
    name: str
    platform: str  # "mac", "windows", "linux"
    ip: str
    port: int
    screens: List[ScreenInfo]
    is_server: bool = False


@dataclass
class MouseMoveEvent:
    """Mouse movement event"""
    x: int
    y: int
    absolute: bool = True  # False for relative/delta movement


@dataclass
class MouseButtonEvent:
    """Mouse button event"""
    button: str  # "left", "right", "middle"
    action: str  # "down", "up", "click", "double_click"
    x: int = 0
    y: int = 0


@dataclass
class MouseScrollEvent:
    """Mouse scroll event"""
    delta_x: int
    delta_y: int


@dataclass
class KeyEvent:
    """Keyboard event"""
    key_code: int  # Virtual key code
    action: str  # "down", "up"
    modifiers: Dict[str, bool] = None  # shift, ctrl, alt, cmd/win
    
    def __post_init__(self):
        if self.modifiers is None:
            self.modifiers = {"shift": False, "ctrl": False, "alt": False, "meta": False}


@dataclass
class ClipboardData:
    """Clipboard content"""
    content_type: str  # "text", "image", "files"
    data: str  # For text: the string, for others: base64 encoded
    source: str  # Computer ID that originated the copy


@dataclass
class ScreenLayout:
    """Layout of screens across all computers"""
    computers: List[ComputerInfo]
    edges: Dict[str, str]  # "computer_id:screen_id:edge" -> "computer_id:screen_id"


class Message:
    """Base message class"""
    
    def __init__(self, msg_type: MessageType, payload: Any = None):
        self.type = msg_type
        self.payload = payload
    
    def to_json(self) -> str:
        data = {
            "type": self.type.value,
            "payload": self._serialize_payload()
        }
        return json.dumps(data) + "\n"
    
    def _serialize_payload(self):
        if self.payload is None:
            return None
        if hasattr(self.payload, '__dataclass_fields__'):
            return asdict(self.payload)
        if isinstance(self.payload, dict):
            return self.payload
        if isinstance(self.payload, list):
            return [asdict(p) if hasattr(p, '__dataclass_fields__') else p for p in self.payload]
        return self.payload
    
    @classmethod
    def from_json(cls, json_str: str) -> 'Message':
        data = json.loads(json_str.strip())
        msg_type = MessageType(data["type"])
        return cls(msg_type, data.get("payload"))
    
    def __repr__(self):
        return f"Message({self.type.value}, {self.payload})"


# Key mapping between Mac and Windows
KEY_MAP_MAC_TO_WIN = {
    # Modifiers
    55: 91,   # Cmd -> Win
    56: 16,   # Shift
    58: 18,   # Alt/Option
    59: 17,   # Ctrl
    
    # Arrow keys
    123: 37,  # Left
    124: 39,  # Right
    125: 40,  # Down
    126: 38,  # Up
    
    # Function keys
    122: 112,  # F1
    120: 113,  # F2
    99: 114,   # F3
    118: 115,  # F4
    96: 116,   # F5
    97: 117,   # F6
    98: 118,   # F7
    100: 119,  # F8
    101: 120,  # F9
    109: 121,  # F10
    103: 122,  # F11
    111: 123,  # F12
    
    # Special keys
    36: 13,   # Return -> Enter
    48: 9,    # Tab
    51: 8,    # Delete -> Backspace
    53: 27,   # Escape
    49: 32,   # Space
    117: 46,  # Forward Delete -> Delete
}

KEY_MAP_WIN_TO_MAC = {v: k for k, v in KEY_MAP_MAC_TO_WIN.items()}


def translate_key(key_code: int, from_platform: str, to_platform: str) -> int:
    """Translate key code between platforms"""
    if from_platform == to_platform:
        return key_code
    
    if from_platform == "mac" and to_platform == "windows":
        return KEY_MAP_MAC_TO_WIN.get(key_code, key_code)
    elif from_platform == "windows" and to_platform == "mac":
        return KEY_MAP_WIN_TO_MAC.get(key_code, key_code)
    
    return key_code

File: WindowsApp/MacWinControlClient/Protocol.cs
using System;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace MacWinControlClient
{
    /// <summary>
    /// Message types matching Python protocol.py
    /// </summary>
    public enum MessageType
    {
        Hello,
        ModeSwitch,
        MouseMove,
        MouseClick,
        MouseScroll,
        KeyDown,
        KeyUp,
        ClipboardSync,
        ClipboardRequest,
        ScreenInfo,
        Heartbeat,
        Error
    }

    /// <summary>
    /// Base message class for all protocol messages
    /// </summary>
    public class ProtocolMessage
    {
        [JsonPropertyName("type")]
        public string Type { get; set; } = "";
        
        [JsonPropertyName("timestamp")]
        public double Timestamp { get; set; }
        
        [JsonPropertyName("source")]
        public string Source { get; set; } = "windows";
        
        [JsonExtensionData]
        public Dictionary<string, JsonElement>? ExtensionData { get; set; }
    }

    /// <summary>
    /// Hello message for handshake
    /// </summary>
    public class HelloMessage : ProtocolMessage
    {
        [JsonPropertyName("client_name")]
        public string ClientName { get; set; } = "";
        
        [JsonPropertyName("version")]
        public string Version { get; set; } = "2.0.0";
        
        [JsonPropertyName("capabilities")]
        public List<string> Capabilities { get; set; } = new();
        
        public HelloMessage()
        {
            Type = "hello";
        }
    }

    /// <summary>
    /// Mode switch message
    /// </summary>
    public class ModeSwitchMessage : ProtocolMessage
    {
        [JsonPropertyName("active_device")]
        public string ActiveDevice { get; set; } = "";
        
        [JsonPropertyName("edge")]
        public string Edge { get; set; } = "";
        
        public ModeSwitchMessage()
        {
            Type = "mode_switch";
        }
    }

    /// <summary>
    /// Mouse move message
    /// </summary>
    public class MouseMoveMessage : ProtocolMessage
    {
        [JsonPropertyName("x")]
        public double X { get; set; }
        
        [JsonPropertyName("y")]
        public double Y { get; set; }
        
        [JsonPropertyName("dx")]
        public double Dx { get; set; }
        
        [JsonPropertyName("dy")]
        public double Dy { get; set; }
        
        [JsonPropertyName("relative")]
        public bool Relative { get; set; }
        
        public MouseMoveMessage()
        {
            Type = "mouse_move";
        }
    }

    /// <summary>
    /// Mouse click message
    /// </summary>
    public class MouseClickMessage : ProtocolMessage
    {
        [JsonPropertyName("button")]
        public string Button { get; set; } = "left";
        
        [JsonPropertyName("pressed")]
        public bool Pressed { get; set; }
        
        public MouseClickMessage()
        {
            Type = "mouse_click";
        }
    }

    /// <summary>
    /// Mouse scroll message
    /// </summary>
    public class MouseScrollMessage : ProtocolMessage
    {
        [JsonPropertyName("dx")]
        public double Dx { get; set; }
        
        [JsonPropertyName("dy")]
        public double Dy { get; set; }
        
        public MouseScrollMessage()
        {
            Type = "mouse_scroll";
        }
    }

    /// <summary>
    /// Key event message
    /// </summary>
    public class KeyMessage : ProtocolMessage
    {
        [JsonPropertyName("key_code")]
        public int KeyCode { get; set; }
        
        [JsonPropertyName("key_name")]
        public string KeyName { get; set; } = "";
        
        [JsonPropertyName("modifiers")]
        public List<string> Modifiers { get; set; } = new();
        
        public KeyMessage(bool isDown = true)
        {
            Type = isDown ? "key_down" : "key_up";
        }
    }

    /// <summary>
    /// Clipboard sync message
    /// </summary>
    public class ClipboardMessage : ProtocolMessage
    {
        [JsonPropertyName("content_type")]
        public string ContentType { get; set; } = "text";
        
        [JsonPropertyName("content")]
        public string Content { get; set; } = "";
        
        public ClipboardMessage()
        {
            Type = "clipboard_sync";
        }
    }

    /// <summary>
    /// Screen information message
    /// </summary>
    public class ScreenInfoMessage : ProtocolMessage
    {
        [JsonPropertyName("screens")]
        public List<ScreenInfo> Screens { get; set; } = new();
        
        [JsonPropertyName("primary_index")]
        public int PrimaryIndex { get; set; }
        
        public ScreenInfoMessage()
        {
            Type = "screen_info";
        }
    }

    /// <summary>
    /// Screen information structure
    /// </summary>
    public class ScreenInfo
    {
        [JsonPropertyName("name")]
        public string Name { get; set; } = "";
        
        [JsonPropertyName("x")]
        public int X { get; set; }
        
        [JsonPropertyName("y")]
        public int Y { get; set; }
        
        [JsonPropertyName("width")]
        public int Width { get; set; }
        
        [JsonPropertyName("height")]
        public int Height { get; set; }
        
        [JsonPropertyName("is_primary")]
        public bool IsPrimary { get; set; }
    }

    /// <summary>
    /// Protocol helper class for message serialization
    /// </summary>
    public static class Protocol
    {
        public const string VERSION = "2.0.0";
        
        private static readonly JsonSerializerOptions JsonOptions = new()
        {
            PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
            WriteIndented = false
        };

        public static string ToJson(ProtocolMessage message)
        {
            message.Timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;
            return JsonSerializer.Serialize(message, message.GetType(), JsonOptions);
        }

        public static ProtocolMessage? FromJson(string json)
        {
            try
            {
                // First parse to get the type
                var doc = JsonDocument.Parse(json);
                var type = doc.RootElement.GetProperty("type").GetString();

                return type switch
                {
                    "hello" => JsonSerializer.Deserialize<HelloMessage>(json, JsonOptions),
                    "mode_switch" => JsonSerializer.Deserialize<ModeSwitchMessage>(json, JsonOptions),
                    "mouse_move" => JsonSerializer.Deserialize<MouseMoveMessage>(json, JsonOptions),
                    "mouse_click" => JsonSerializer.Deserialize<MouseClickMessage>(json, JsonOptions),
                    "mouse_scroll" => JsonSerializer.Deserialize<MouseScrollMessage>(json, JsonOptions),
                    "key_down" or "key_up" => JsonSerializer.Deserialize<KeyMessage>(json, JsonOptions),
                    "clipboard_sync" => JsonSerializer.Deserialize<ClipboardMessage>(json, JsonOptions),
                    "screen_info" => JsonSerializer.Deserialize<ScreenInfoMessage>(json, JsonOptions),
                    _ => JsonSerializer.Deserialize<ProtocolMessage>(json, JsonOptions)
                };
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[Protocol] Parse error: {ex.Message}");
                return null;
            }
        }

        public static HelloMessage CreateHello(string clientName, List<string>? capabilities = null)
        {
            return new HelloMessage
            {
                ClientName = clientName,
                Version = VERSION,
                Capabilities = capabilities ?? new List<string> { "mouse", "keyboard", "clipboard" }
            };
        }

        public static ModeSwitchMessage CreateModeSwitch(string activeDevice, string edge = "")
        {
            return new ModeSwitchMessage
            {
                ActiveDevice = activeDevice,
                Edge = edge
            };
        }

        public static MouseMoveMessage CreateMouseMove(double x, double y, double dx = 0, double dy = 0, bool relative = false)
        {
            return new MouseMoveMessage
            {
                X = x,
                Y = y,
                Dx = dx,
                Dy = dy,
                Relative = relative
            };
        }

        public static ClipboardMessage CreateClipboardSync(string content, string contentType = "text")
        {
            return new ClipboardMessage
            {
                Content = content,
                ContentType = contentType
            };
        }

        public static ScreenInfoMessage CreateScreenInfo(List<ScreenInfo> screens, int primaryIndex = 0)
        {
            return new ScreenInfoMessage
            {
                Screens = screens,
                PrimaryIndex = primaryIndex
            };
        }
    }
}
